internal/corazawaf/transaction.go

```diff
@@ -748,8 +749,75 @@ func (tx *Transaction) ProcessRequestHeaders() *types.Interruption {
 	return tx.interruption
 }
 
-func (tx *Transaction) RequestBodyWriter() io.Writer {
-	return tx.RequestBodyBuffer
+type Lenger interface {
+	Len() int
+}
+
+func (tx *Transaction) WriteRequestBody(b []byte) (*types.Interruption, int, error) {
+	if tx.RuleEngine == types.RuleEngineOff {
+		return nil, 0, nil
+	}
+
+	writingBytes := int64(len(b))
+	if tx.requestBodyBuffer.length+writingBytes >= tx.RequestBodyLimit {
+		tx.variables.inboundErrorData.Set("1")
+		if tx.WAF.RequestBodyLimitAction == types.RequestBodyLimitActionReject {
+			// We interrupt this transaction in case RequestBodyLimitAction is Reject
+			tx.interruption = &types.Interruption{
+				Status: 403,
+				Action: "deny",
+			}
+			return tx.interruption, 0, nil
+		}
+
+		if tx.WAF.RequestBodyLimitAction == types.RequestBodyLimitActionProcessPartial {
+			writingBytes = tx.RequestBodyLimit - tx.requestBodyBuffer.length
+		}
+	}
+
+	w, err := tx.requestBodyBuffer.Write(b[:writingBytes])
+	if err != nil {
+		return nil, 0, err
+	}
+
+	return nil, int(w), nil
+}
+
+// ProcessRequestBodySize sets the maximum number of bytes that can be read from the
+// request body.
+func (tx *Transaction) WriteRequestBodyFrom(r io.Reader) (*types.Interruption, int, error) {
+	if tx.RuleEngine == types.RuleEngineOff {
+		return nil, 0, nil
+	}
+
+	var writingBytes int64
+	if l, ok := r.(Lenger); ok {
+		writingBytes = int64(l.Len())
+		if tx.requestBodyBuffer.length+writingBytes >= tx.RequestBodyLimit {
+			tx.variables.inboundErrorData.Set("1")
+			if tx.WAF.RequestBodyLimitAction == types.RequestBodyLimitActionReject {
+				// We interrupt this transaction in case RequestBodyLimitAction is Reject
+				tx.interruption = &types.Interruption{
+					Status: 403,
+					Action: "deny",
+				}
+				return tx.interruption, 0, nil
+			}
+
+			if tx.WAF.RequestBodyLimitAction == types.RequestBodyLimitActionProcessPartial {
```

Comment:
I don't think we should trigger ProcessRequestBody if not called by user as they might chose to add more data after body appending (or not). In any case having non explicit behaviours is usually a sourcer of bugs.

> adding a check of returned written bytes inside the middleware (connectors): if not all bytes have been written and an interruption has not been raised (reject interruption), it means that we are ready to partially process the body. So we can call the process and, from now on, just let the body be sent upstream

This is actually a good point, looking at the written bytes is one option or adding another returning value is another option but I'd rather document the written bytes behaviour.

Reply:
We're returning `Interruption` from this method now so definitely think we should process the request body - if the user adds more bytes before processing an interrupted request, that's wasted processing and would itself be a bug in the connector anyways.

We need to make the processing idempotent for it, which we should do anyways to allow phase catch up as described in https://github.com/corazawaf/coraza-proxy-wasm/pull/128#discussion_r1062070317

Reply:
> if the user adds more bytes before processing an interrupted request, that's wasted processing and would itself be a bug in the connector anyways.

Do you mean to call the ProcessRequestBody? If that is the case I don't think we should because the result will be the same, still agree with the idempotency, we just need to decide if after an interruption is raised all the calls to `ProcessX` methods should act as noop and return the existing interruption (misleading phase as you would receive an interruption on ProcessResponseHeaders when if ignored the interruption on appending body) or panic (connectors would avoid this as programatic error). I will raise this in a yet another PR.

Reply:
Yes I think we should call `ProcessRequestBody` - the result isn't the same since that would actually evaluate phase 2 rules, not just a body limit right? With the idempotency changes including #574 there shouldn't be any problem and it will keep request processing logic in the connectors the simplest / most uniform for dealing with ProcessPartial.